src/passes/Inlining.cpp

```diff
@@ -62,6 +64,22 @@ struct FunctionInfo {
     hasTryDelegate = false;
     usedGlobally = false;
     uninlineable = false;
+
+    // Normally we do not inline functions that do heavy work, as if they run
+    // for a long time then getting rid of the call overhead to them has little
+    // benefit (an exception is where we can reduce the amount of work by using
+    // knowledge about the parameters passed to them - we should do more there
+    // TODO). However, some functions may skip their heavy work, such as
+    //
+    // function foo(x) {
+    //   if (!x) return;
+    //   heavyWork();
+    // }
+    //
+    // That function *is* worth inlining, as the heavy work is conditional, and
+    // if in practice it is avoided, then avoiding the call overhead there can
```

Comment:
IIUC, there is some risk here for situations where the heavy work does happen in practice. Are we concerned about this at all?

Reply:
Risk in terms of speed? That should be low, I think. The main risk here is code size cost, which we at least know up front. The potential speedup is something we hope for, but that's about it.

I think I think this is a plausible rule, and it does look for very very minimal unavoidable work - almost zero - so it should not apply excessively. But we could make it even more specific if we look for some kind of indication that the skippable code is unlikely. If we had branch-hinting that would be good; otherwise, perhaps throwing an exception is such a hint.

Reply:
Right, I was thinking of code size risk. But you explained that code size is tracked separately, so this seems fine to me. I agree that there are good heuristics we could use to guess at branch probability like throwing or trapping. I would expect us to add those as part of some future work to support branch-hinting.